packages/react-slider/src/components/Slider/useSliderState.tsx

```diff
@@ -0,0 +1,237 @@
+ */
+const getPercent = (value: number, min: number, max: number) => {
+  return max === min ? 0 : ((value - min) / (max - min)) * 100;
+};
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+const on = (element: Element, eventName: string, callback: (ev: any) => void) => {
+  element.addEventListener(eventName, callback);
+  return () => element.removeEventListener(eventName, callback);
+};
+
+export const useSliderState = (state: Pick<SliderState, keyof SliderCommon | keyof SliderSlots | 'as' | 'ref'>) => {
+  const {
+    as = 'div',
+    value,
+    defaultValue = 0,
+    min = 0,
+    max = 10,
+    step = 1,
+    ariaValueText,
+    onChange,
+    onPointerDown: onPointerDownCallback,
+    onKeyDown: onKeyDownCallback,
+  } = state;
+
+  const [currentValue, setCurrentValue] = useControllableState({
+    state: value && clamp(value, min, max),
+    defaultState: clamp(defaultValue, min, max),
+    initialState: 0,
+  });
+
+  const railRef = React.useRef<HTMLDivElement>(null);
+  const thumbRef = React.useRef<HTMLDivElement>(null);
+  const disposables = React.useRef<(() => void)[]>([]);
+  const onChangeCallback = React.useRef(onChange);
+  const id = useId('slider-', state.id);
+
+  /**
+   * Updates the `currentValue` to the new `incomingValue` and clamps it.
+   *
+   * @param ev
+   * @param incomingValue
+   */
+  const updateValue = React.useCallback(
+    (incomingValue: number, ev): void => {
+      const clampedValue = clamp(incomingValue, min, max);
+
+      if (clampedValue !== min && clampedValue !== max) {
+        ev.stopPropagation();
+        if (ev.type === 'keydown') {
+          ev.preventDefault();
+        }
+      }
+
+      if (onChange && onChangeCallback.current) {
+        onChangeCallback.current(clampedValue, ev);
+      }
+
+      setCurrentValue(clampedValue);
+    },
+    [max, min, onChange, setCurrentValue],
+  );
+
+  /**
+   * Calculates the `step` position based off of a `Mouse` or `Touch` event.
+   */
+  const calculateSteps = React.useCallback(
+    (ev: React.PointerEvent<HTMLDivElement>): number => {
+      const currentBounds = railRef?.current?.getBoundingClientRect();
+      const size = currentBounds?.width || 0;
+      const position = currentBounds?.left || 0;
+
+      const totalSteps = (max - min) / step;
+      const stepLength = size / totalSteps;
+      const thumbPosition = ev.clientX;
+      const distance = thumbPosition - position;
+      return distance / stepLength;
+    },
+    [max, min, step],
+  );
+
+  const onPointerMove = React.useCallback(
+    (ev: React.PointerEvent<HTMLDivElement>): void => {
+      if (step !== 1) {
+        updateValue(Math.round((min + step * calculateSteps(ev)) / step) * step, ev);
+      } else {
+        updateValue(min + step * calculateSteps(ev), ev);
+      }
+    },
+    [calculateSteps, min, step, updateValue],
+  );
+
+  const onPointerUp = (): void => {
+    disposables.current.forEach(dispose => dispose());
```

Comment:
When you mouse down on the track, focus is not moved to the thumb but rather the body. Focus should be moved to the appropriate thumb.

This should be down on pointer down, but could also be done on pointer up. In testing, this definitely worked with pointer up, but something is preventing pointerdown from moving focus to thumb. Maybe just fix pointer up for now to move focus, but investigate if that logic can be moved.